library/src/scala/util/CommandLineParser.scala

```diff
@@ -0,0 +1,44 @@
+package scala.util
+
+/** A utility object to support command line parsing for @main methods */
+object CommandLineParser {
```

Comment:
I understand the goal of simplifying the way to implement application entry points. In practice, most of the users just write a `def main(args: Array[String]): Unit` method, and manually parse the application arguments. For these cases, this PR definitely improves the developer experience.

That being said, by introducing a way to parse the program arguments in the compiler, we will also establish a “standard” way of doing that. And, I’m a little bit worried about the introduction of the `CommandLineParser` and `FromString` abstractions. For now, they are simple but I think we can already anticipate the following requests soon: can we also support optional parameters? Can we support named parameters? How could we customize the error message in case of invalid parameters? Can we print all the error messages instead of stopping at the first one? Should we use `Either` instead of relying on exceptions to model parsing failures? All these requests are legitimate, in my opinion, but if we try to fulfill them the size of the `CommandLineParser` and `FromString` modules will be 10x bigger. Then, the question will be “should we keep that thing in the compiler, or should we move it to the library land (like scopt, or optparse-applicative)?”. We can make the decision of keeping `CommandLineParser` simple and not honor these requests, “by design”, but then I’m worried about introducing a “standard” way of parsing command-line arguments which does not scale to advanced use cases.

Unfortunately, I don’t have a simple idea of how to decouple `@main` expansion from `CommandLineParser` (so that users could plug in the parsing + reporting logic they want). I still think the most flexible way to go is to require users to forward a plain old `main` method to whatever they want:

~~~ scala
def happyBirthday(age: Int, name: String, others: String*): Unit = …

def main(args: Array[String]): Unit = {
  val happyBirthdayParser = … // some code that relies on a library for arguments parsing
  happyBirthdayParser.run(args)(happyBirthday)
}
~~~

But I agree that this approach is *not* beginner-friendly…

Reply:
I believe we can stay firm and keep `CommandLineParser` simple. If one wants something else, one can perfectly well define a traditional main method, or define the entry point like this:
```scala
@main def f(args: String*) = ...
```